x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authz/RBACEngine.java

```diff
@@ -880,19 +881,14 @@ static AuthorizedIndices resolveAuthorizedIndicesFromRole(
         }, name -> {
             final IndexAbstraction indexAbstraction = lookup.get(name);
             if (indexAbstraction == null) {
-                return false;
-            }
-            if (includeDataStreams) {
+                // missing but authorized resources should be handled downstream in the action handler, not here in the Security filter
+                return predicate.testMissingResource(name);
+            } else {
                 // We check the parent data stream first if there is one. For testing requested indices, this is most likely
                 // more efficient than checking the index name first because we recommend grant privileges over data stream
                 // instead of backing indices.
-                if (indexAbstraction.getParentDataStream() != null && predicate.test(indexAbstraction.getParentDataStream())) {
-                    return true;
-                } else {
-                    return predicate.test(indexAbstraction);
-                }
-            } else {
-                return indexAbstraction.getType() != IndexAbstraction.Type.DATA_STREAM && predicate.test(indexAbstraction);
+                return (indexAbstraction.getParentDataStream() != null && predicate.test(indexAbstraction.getParentDataStream()))
+                    || predicate.test(indexAbstraction);
             }
```

Comment:
This is the core of the change proposed by this PR.

This predicate here is used to remove explicit names when rewriting the request if `ignore_unavailable` is `true`, see: https://github.com/elastic/elasticsearch/blob/844bb4cee573b4eb2b66fd2253d364ae7cf935fa/server/src/main/java/org/elasticsearch/cluster/metadata/IndexAbstractionResolver.java#L112

After this expression rewrite, the request goes into the authz stage, where every expression item is checked again: 
https://github.com/elastic/elasticsearch/blob/844bb4cee573b4eb2b66fd2253d364ae7cf935fa/x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/security/authz/permission/IndicesPermission.java#L510

The intent is that the test here, ie:
``` 
final IndexAbstraction indexAbstraction = lookup.get(name);
if (indexAbstraction == null) {
    // missing but authorized resources should be handled downstream in the action handler, not here in the Security filter
    return predicate.testMissingResource(name);
} else {
    // We check the parent data stream first if there is one. For testing requested indices, this is most likely
    // more efficient than checking the index name first because we recommend grant privileges over data stream
    // instead of backing indices.
    return (indexAbstraction.getParentDataStream() != null && predicate.test(indexAbstraction.getParentDataStream()))
        || predicate.test(indexAbstraction);
}
```
be equivalent to the test in `IndicesPermission#isActionGranted`, such that **only unauthorized names be removed when rewriting the expression with ignore_unavailable=true**.

Reply:
Can we unify the loigc of `test` and `testMissingResource` into a single method? It is better that callers do not have to worry about deciding on which method to use. It should be possible if the predicate takes an object that encapsulate both `name` and `indexAbstraction`, something similar to `IndicesPermission#IndexResource`. 

I think ideally it would be great to reuse the existing `IndexResource` and unify the logic between this method and `isActionGranted`. It will likely help understand this part of code. But it does not seem to be an easy job. So we can tackle it later gradually. But I think we can avoid having both `test` and `testMissingResource` as suggest earlier. It also helps the `IsResourceAuthorizedPredicate` to be just a simple `Predicate` and avoid having to implement its own `and` method.

Reply:
There's a subtle problem with making the `IsResourceAuthorizedPredicate` take an `IndexResource` parameter.
The problem is that, when the resource exists, it matters if the "is resource authorized" predicate is called in the context of wildcard expansion, or in the context of checking explicit names.

The difference is that the predicate in the context of checking explicit names first tests on the parent datastream (if applicable) and then it tests on the resource itself. This is not the case for the predicate in the wildcard expansion context.

So, in essence, the `IsResourceAuthorizedPredicate` would still have to implement two slightly different test methods.

Not exactly what you asked for, but I could be hidding this whole
```
final IndexAbstraction indexAbstraction = lookup.get(name);
if (indexAbstraction == null) {
    // missing but authorized resources should be handled downstream in the action handler, not here in the Security filter
    return predicate.testMissingResource(name);
} else {
    // We check the parent data stream first if there is one. For testing requested indices, this is most likely
    // more efficient than checking the index name first because we recommend grant privileges over data stream
    // instead of backing indices.
    return (indexAbstraction.getParentDataStream() != null && predicate.test(indexAbstraction.getParentDataStream()))
        || predicate.test(indexAbstraction);
}
```
block inside a method in the `IsResourceAuthorizedPredicate` class, something of the sorts of `#testExplicitName`. This should be equivalent (and someday replace) the `isActionGranted` method.
What do you think?

Reply:
My preference is to avoid having a second test method for `IsResourceAuthorizedPredicate`. It is surprising that a predicate has more than one test method and you need some effort in deciding which one to use. 

Based on your comment here https://github.com/elastic/elasticsearch/pull/92625#discussion_r1071358409

Can we make `IsResourceAuthorizedPredicate` implement `BiPredicate` if you prefer not having a `IndexResource` like class?

Reply:
As we've discussed earlier today, I've reimplemented the `IsResourceAuthorizedPredicate` as a `BiPredicate<String, IndexAbstraction>`, see [878ba9e](https://github.com/elastic/elasticsearch/pull/92625/commits/878ba9e21f5d07ae3fd4a060617b31b532d78d47) :
```
    public static class IsResourceAuthorizedPredicate implements BiPredicate<String, IndexAbstraction> {

        private final BiPredicate<String, IndexAbstraction> biPredicate;

        // public for tests
        public IsResourceAuthorizedPredicate(StringMatcher resourceNameMatcher, StringMatcher additionalNonDatastreamNameMatcher) {
            this((String name, @Nullable IndexAbstraction indexAbstraction) -> {
                assert indexAbstraction == null || name.equals(indexAbstraction.getName());
                return resourceNameMatcher.test(name)
                    || (isPartOfDatastream(indexAbstraction) == false && additionalNonDatastreamNameMatcher.test(name));
            });
        }

        private IsResourceAuthorizedPredicate(BiPredicate<String, IndexAbstraction> biPredicate) {
            this.biPredicate = biPredicate;
        }

        public final IsResourceAuthorizedPredicate and(IsResourceAuthorizedPredicate other) {
            return new IsResourceAuthorizedPredicate(this.biPredicate.and(other.biPredicate));
        }

        public final boolean test(IndexAbstraction indexAbstraction) {
            return test(indexAbstraction.getName(), indexAbstraction);
        }

        @Override
        public boolean test(String name, @Nullable IndexAbstraction indexAbstraction) {
            return biPredicate.test(name, indexAbstraction);
        }

        private static boolean isPartOfDatastream(IndexAbstraction indexAbstraction) {
            return indexAbstraction != null
                && (indexAbstraction.getType() == IndexAbstraction.Type.DATA_STREAM || indexAbstraction.getParentDataStream() != null);
        }
    }
```
For convenience, I also added a `test(IndexAbstraction indexAbstraction)` in addition to `test(String name, @Nullable IndexAbstraction indexAbstraction)`, to avoid the `test(index.getName(), index)` pattern.

Reply:
However, I'm hesitant to change the predicate further, in order to move the logic that tests the parent datastream of a backing index. I still can't see any serious issues, but I very much prefer that such a change be in a separate PR.